src/Behat/Gherkin/Node/FeatureNode.php

```diff
@@ -64,7 +66,7 @@ class FeatureNode implements KeywordNodeInterface, TaggedNodeInterface
      * @param ScenarioInterface[] $scenarios
      * @param string              $keyword
      * @param string              $language
-     * @param null|string         $file
+     * @param null|string         $file        The absolute path to the feature file.
```

Comment:
What reason do we have for this to be allowed as null? Since we've got a (minor) API break here, cleaning this up would be a nice improvement.

Reply:
Good suggestion, I will look into this. I know that a number of tests are calling this without a file being passed, but this might just be _because_ it is an optional parameter.

From the first look it seems like this should be required indeed. The line number is also required, and what point is it to have a line number if there is no file?

I'll have a look at the [most popular packages that require Gherkin](https://packagist.org/packages/behat/gherkin/dependents?page=1) and see if they have any cases where the class is instantiated without a file. And also, let's throw a deprecation warning for the moment, and actually drop it in a major update.

Reply:
I started working on this but there is actually a valid use case for not having a `$file`. You can call `Parser::parse($input)` to parse a string containing a Gherkin feature that is not saved in a file. This will end up calling `Parser::parseFeature()` which instantiates a `FeatureNode` without passing a `$file`. This also explains why the `$line` is required but the `$file` is not; even when passing a string directly, this will consist of multiple lines, so you can set a meaningful line number without having a file.

Ref. https://github.com/Behat/Gherkin/blob/master/src/Behat/Gherkin/Parser.php#L62